Don't `@inbounds` AbstractArray's iterate method; optimize `checkbounds` instead

mbauman opened the pull request:
Split off from #58785, this simplifies `iterate` and removes the `@inbounds` call that was added in https://github.com/JuliaLang/julia/pull/58635.  It achieves the same (or better!) performance, however, by targeting optimizations in `checkbounds` and — in particular — the construction of a linear `eachindex` (against which the bounds are checked).

giordano committed:
no inbounds on AbstractArray iterate and axe Array/Memory methods

mbauman committed:
avoid branch in OrdinalRange length

mbauman commented:
Using @jishnub's benchmark:

Nightly @ f61c640aaf2 (which has the `@inbounds`):

```julia
julia> using BenchmarkTools
Precompiling BenchmarkTools finished.
  8 dependencies successfully precompiled in 12 seconds. 9 already precompiled.

julia> using LinearAlgebra

julia> A = rand(1000,1000); v2 = view(A, 1:2:lastindex(A));

julia> @btime norm(Iterators.map(splat(-), zip($v2, $v2)));
  1.941 ms (0 allocations: 0 bytes)
```

vs. [b09514e](https://github.com/JuliaLang/julia/pull/58793/commits/b09514e4632088cc0d7819c3ef0a2d4e7e62eeaf):

```Julia
julia> @btime norm(Iterators.map(splat(-), zip($v2, $v2)));
  312.167 μs (0 allocations: 0 bytes)
```

oscardssmith commented on base/abstractarray.jl:
```diff
@@ -1237,15 +1238,15 @@ oneunit(x::AbstractMatrix{T}) where {T} = _one(oneunit(T), x)
-iterate(A::AbstractArray, state = iterate_starting_state(A)) = _iterate(A, state)
-function _iterate(A::AbstractArray, state::Tuple)
+@inline iterate(A::AbstractArray, state = iterate_starting_state(A)) = _iterate(A, state)
+@inline function _iterate(A::AbstractArray, state::Tuple)
     y = iterate(state...)
     y === nothing && return nothing
     A[y[1]], (state[1], tail(y)...)
 end
-function _iterate(A::AbstractArray, state::Integer)
+@inline function _iterate(A::AbstractArray, state::Integer)
```
are these functions really so big that they aren't being inlined automatically?

mbauman commented:
@nanosoldier `runbenchmarks()`

giordano commented:
Test failures (quite a lot) look relevant 😢

mbauman committed:
fix SubArray axes1

mbauman committed:
implement optimized checkbounds for FastSubArray

mbauman committed:
Remove inbounds on CodeUnits, too

mbauman committed:
add tests

mbauman committed:
fixup! fix SubArray axes1

mbauman committed:
remove bugged SubArray checkbounds method

This is not valid because it only checks the resulting index into the parent (which we explicitly say we do not check) but skips the checks into the indices (which are the important ones!)

mbauman committed:
Merge remote-tracking branch 'origin/master' into mb+mg/array-iteration

mbauman committed:
CodeUnits: implement checkbounds and fallback for iterate

mbauman committed:
do not test for unreachable with check-bounds=yes

theoretically this is not guaranteed -- indeed an unreachable branch is not currently present for CodeUnits

mbauman commented on base/strings/basic.jl:
```diff
@@ -797,7 +797,7 @@ size(s::CodeUnits) = (length(s),)
 @propagate_inbounds getindex(s::CodeUnits, i::Int) = codeunit(s.s, i)
 IndexStyle(::Type{<:CodeUnits}) = IndexLinear()
-@inline iterate(s::CodeUnits, i=1) = checkbounds(Bool, s, i) ? (@inbounds s[i], i + 1) : nothing
+checkbounds(::Type{Bool}, s::CodeUnits, i) = checkbounds(Bool, s.s, i)
```
```suggestion
checkbounds(::Type{Bool}, s::CodeUnits, i::Integer) = checkbounds(Bool, s.s, i)
```

mbauman committed:
fix ambiguity and more properly constrain checkbounds for CodeUnits

JeffBezanson commented:
Very good PR. Benchmarks are an interesting mixed bag of slowdowns and infinite speedups :) Not sure what to do about it.

mbauman commented:
🤷 

@nanosoldier `runbenchmarks(ALL, vs=":master")`

vtjnash commented:
I don't think nanosoldier knows how to distribute `!` over `||`, and `[` means a single test that matches all of those tags in order

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/58793#issuecomment-3013507831) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4b09ab3_vs_4ae3f5e/report.md).

mbauman commented:
Just for posterity, as I understand it, there are three separate considerations — each of which help increase the chance that the compiler does something smart here:

* Ensuring `length` inlines (here by making a branch do math instead)
* The iteration implementation here takes the form `checkbounds(Bool, x, i) ? (x[i], i+1) : nothing`... and `x[i]` _itself_ should have the exact same `checkbounds(Bool, x, i)` predicate inside it.  Making sure those two predicates match _exactly_ increases the odds that the compiler skips the inner branch entirely.  (That's what's behind the `CodeUnits` change)
* Making sure the `checkbounds` test itself is linear (purely adds and muls) helps LLVM track repeated `checkbounds` inside a for loop and hoist it out entirely. (That's what moving to `unchecked_oneto` does)

giordano reviewed (approved):
I'm not deeply familiar with this part of the code, but changes look sensible, the diff in `base` (excluding `test`) is net negative, we get new tests and more consistently better performance, looks like a win-win situation.

mbauman commented on base/range.jl:
```diff
@@ -810,10 +810,8 @@ let bigints = Union{Int, UInt, Int64, UInt64, Int128, UInt128},
-            a = div(unsigned(-diff), -s) % typeof(diff)
         else
-            a = div(unsigned(diff), s) % typeof(diff)
+            a = div(unsigned(sign(s)*diff), abs(s)) % typeof(diff)
```
Ah, that branch is unfortunately still necessary for (at least) the `s == -s` case:

```
julia> r = typemax(Int):typemin(Int):typemin(Int)
9223372036854775807:-9223372036854775808:-1

julia> s = step(r)
-9223372036854775808

julia> diff = last(r) - first(r)
-9223372036854775808

julia> div(diff, s) % typeof(diff)
1

julia> div(unsigned(flipsign(diff, s)), abs(s)) % typeof(diff)
-1
```

mbauman replied:
Ooooh, perhaps it could work with:

```
julia> div(unsigned(flipsign(diff, s)), unsigned(abs(s)))
0x0000000000000001
```

mbauman committed:
use flipsign instead of multiplying sign

Co-authored-by: N5N3 <2642243996@qq.com>

mbauman committed:
defer length(::OrdinalRange) changes; just inline it for now

mbauman commented:
OK, given that the `length` changes here are relatively ancillary to the main objective around iteration, I've split those off into #58864, and just mark them `@inline` for now.  Given the fundamental importance of the length of a range, I figure that's prudent.  Just marking them inline gets us _some_ of the advantages in the original benchmark that is very specifically targeting the inlining of this method (and its callers).  That benchmark is now half-way between what I reported in https://github.com/JuliaLang/julia/pull/58793#issuecomment-2997650614:

```Julia
julia> using BenchmarkTools, LinearAlgebra

julia> A = rand(1000,1000); v2 = view(A, 1:2:lastindex(A));

julia> @btime norm(Iterators.map(splat(-), zip($v2, $v2)));
  931.917 μs (0 allocations: 0 bytes)
```
